### ISSUE TITLE
Microsoft.AspNetCore.Identity.EntityFrameworkCore.UserStore violates the constraint of type parameter 'TUser'.

### ISSUE THREAD

**Issue by MaklaCof** [aspnet/Identity#1082]
I am trying to use Id as int (Identity). My version is 1.1
![devenv_2017-01-18_13-09-21](https://cloud.githubusercontent.com/assets/1589573/22063499/6a4c2da4-dd7f-11e6-9a8b-c673a0c1da99.png)

I override all Identity classes:
```
    [Table("Roles")]
    public partial class AppRole : IdentityRole<int, AppUserRole, AppRoleClaim>
    {
    }
    [Table("RoleClaims")]
    public partial class AppRoleClaim : IdentityRoleClaim<int>
    {
    }
    [Table("Users")]
    public partial class AppUser : IdentityUser<int, AppUserClaim, AppUserRole, AppUserLogin>
    {       
    }
    [Table("UserClaims")]
    public partial class AppUserClaim : IdentityUserClaim<int>
    {
    }
    [Table("UserLogins")]
    public partial class AppUserLogin : IdentityUserLogin<int>
    {
        [Key]
        [Required]
        public int Id { get; set; }
    }
    [Table("UserRoles")]
    public partial class AppUserRole : IdentityUserRole<int>
    {
    }
    [Table("UserTokens")]
    public partial class AppUserToken : IdentityUserToken<int>
    {
        [Key]
        [Required]
        public int Id { get; set; }
    }
```
and my context is:
```
    public partial class AppDbContext : IdentityDbContext<AppUser, AppRole, int, AppUserClaim, AppUserRole, AppUserLogin, AppRoleClaim, AppUserToken>
    {

        public AppDbContext() : base() {
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"...");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<AppRole>(entity =>
            {
                entity
                .HasKey(u => u.Id);
                entity.Property(p => p.Id)
                .ValueGeneratedOnAdd();
            });

            modelBuilder.Entity<AppRoleClaim>(entity =>
            {
                entity
                .HasKey(u => u.Id);
                entity.Property(p => p.Id)
                .ValueGeneratedOnAdd();
            });

            modelBuilder.Entity<AppUser>(entity =>
            {
                entity
                .HasKey(u => u.Id);
                entity.Property(p => p.Id)
                .ValueGeneratedOnAdd();
            });

            modelBuilder.Entity<AppUserClaim>(entity =>
            {
                entity
                .HasKey(u => u.Id);
                entity.Property(p => p.Id)
                .ValueGeneratedOnAdd();
            });

            modelBuilder.Entity<AppUserLogin>(entity =>
            {
                entity
                .HasKey(u => u.Id);
                entity.Property(p => p.Id)
                .ValueGeneratedOnAdd();
            });

            modelBuilder.Entity<AppUserRole>(entity =>
            {
                entity
                .HasKey(u => new { u.RoleId, u.UserId });
            });

            modelBuilder.Entity<AppUserToken>(entity =>
            {
                entity
                .HasKey(u => u.Id);
                entity.Property(p => p.Id)
                .ValueGeneratedOnAdd();
            });
        }
    }
```
and startup.cs:
```
                services.AddMvc();

                services.AddEntityFrameworkSqlServer();

                var connection = Configuration["ConnectionStrings"];
                services.AddDbContext<AppDbContext>(options => options.UseSqlServer(connection));

                var res = services
                    .AddIdentity<AppUser, AppRole>(config =>
                    {
                        config.User.RequireUniqueEmail = true;
                        config.Password.RequireNonAlphanumeric = false;
                        config.Cookies.ApplicationCookie.AutomaticChallenge = false;
                    })
                    .AddUserStore<ApplicationUserStore>()
                .AddEntityFrameworkStores<AppDbContext, int>()
                .AddDefaultTokenProviders();
```

I found few web-pages describing this error and all suggest upgrading to 1.1.
Exceptions is raised when debugging in line `.AddEntityFrameworkStores<AppDbContext, int>()`
My complete error displayed in page is:

> TypeLoadException: GenericArguments[0], 'OpPISWeb.Models.AppUser', on 'Microsoft.AspNetCore.Identity.EntityFrameworkCore.UserStore`9[TUser,TRole,TContext,TKey,TUserClaim,TUserRole,TUserLogin,TUserToken,TRoleClaim]' violates the constraint of type parameter 'TUser'.
> System.RuntimeTypeHandle.Instantiate(RuntimeTypeHandle handle, IntPtr* pInst, int numGenericArgs, ObjectHandleOnStack type)
> 
> ArgumentException: GenericArguments[0], 'OpPISWeb.Models.AppUser', on 'Microsoft.AspNetCore.Identity.EntityFrameworkCore.UserStore`4[TUser,TRole,TContext,TKey]' violates the constraint of type 'TUser'.
> System.RuntimeType.ValidateGenericArguments(MemberInfo definition, RuntimeType[] genericArguments, Exception e)
> 
> TypeLoadException: GenericArguments[0], 'OpPISWeb.Models.AppUser', on 'Microsoft.AspNetCore.Identity.EntityFrameworkCore.UserStore`9[TUser,TRole,TContext,TKey,TUserClaim,TUserRole,TUserLogin,TUserToken,TRoleClaim]' violates the constraint of type parameter 'TUser'.
> System.RuntimeTypeHandle.Instantiate(RuntimeTypeHandle handle, IntPtr* pInst, int numGenericArgs, ObjectHandleOnStack type)
> System.RuntimeTypeHandle.Instantiate(Type[] inst)
> System.RuntimeType.MakeGenericType(Type[] instantiation)
> 
> Show raw exception details
> ArgumentException: GenericArguments[0], 'OpPISWeb.Models.AppUser', on 'Microsoft.AspNetCore.Identity.EntityFrameworkCore.UserStore`4[TUser,TRole,TContext,TKey]' violates the constraint of type 'TUser'.
> System.RuntimeType.ValidateGenericArguments(MemberInfo definition, RuntimeType[] genericArguments, Exception e)
> System.RuntimeType.MakeGenericType(Type[] instantiation)
> Microsoft.Extensions.DependencyInjection.IdentityEntityFrameworkBuilderExtensions.GetDefaultServices(Type userType, Type roleType, Type contextType, Type keyType)
> Microsoft.Extensions.DependencyInjection.IdentityEntityFrameworkBuilderExtensions.AddEntityFrameworkStores<TContext, TKey>(IdentityBuilder builder)
> OpPISWeb.Startup.ConfigureServices(IServiceCollection services)
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
> Microsoft.AspNetCore.Hosting.ConventionBasedStartup.ConfigureServices(IServiceCollection services)
> Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureApplicationServices()
> Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()
> 
> Show raw exception details
> .NET Core X64 v4.1.1.0    |   Microsoft.AspNetCore.Hosting version 1.1.0-rtm-22752    |    Microsoft Windows 10.0.14393    |   Need help? Yes please. :)

What did I miss?

**Comment by VahidN** [comment]
When you are customizing ASP.NET Core Identity, you should not use `AddEntityFrameworkStores` anymore. Because it will override all of your previous settings and customization to default Identity services.
First you need to  create new services with the following signatures:
```csharp
public class ApplicationRoleManager:  RoleManager<Role>
public class ApplicationRoleStore :  RoleStore<Role, ApplicationDbContext, int, UserRole, RoleClaim>
public class ApplicationSignInManager :  SignInManager<User>
public class ApplicationUserManager : UserManager<User>
public class ApplicationUserStore : UserStore<User, Role, ApplicationDbContext, int, UserClaim, UserRole, UserLogin, UserToken, RoleClaim>
```
Then redirect built-in services to your custom services:
```csharp
services.AddScoped<UserStore<User, Role, ApplicationDbContext, int, UserClaim, UserRole, UserLogin, UserToken, RoleClaim>, ApplicationUserStore>();
services.AddScoped<UserManager<User>, ApplicationUserManager>();
services.AddScoped<RoleManager<Role>, ApplicationRoleManager>();
services.AddScoped<SignInManager<User>, ApplicationSignInManager>();
services.AddScoped<RoleStore<Role, ApplicationDbContext, int, UserRole, RoleClaim>, ApplicationRoleStore>();
services.AddScoped<IEmailSender, AuthMessageSender>();
services.AddScoped<ISmsSender, AuthMessageSender>();
```
now introduce your custom services:
```csharp
services.AddIdentity<User, Role>(identityOptions =>
            {
             // ...
            }).AddUserStore<ApplicationUserStore>()
              .AddUserManager<ApplicationUserManager>()
              .AddRoleStore<ApplicationRoleStore>()
              .AddRoleManager<ApplicationRoleManager>()
              .AddSignInManager<ApplicationSignInManager>()
              // You **cannot** use .AddEntityFrameworkStores() when you customize everything
              //.AddEntityFrameworkStores<ApplicationDbContext, int>()
              .AddDefaultTokenProviders();
```

**Comment by HaoK** [comment]
In 2.0 AddEntityFrameworkStores was improved to make things easier, the TKey argument was removed and the DbContext is sufficient to 'do the right thing'.  You don't need to actually register the ApplicationXyz services unless you are consuming those services in your own code.  

But AddUser/RoleStore is needed at a minimum for this to work.

**Comment by baseless** [comment]
See the same issue with Guid keys. It starts once you extend beyond TUser, TRole, TKey.

**Comment by MaklaCof** [comment]
Thank you very much.

**Comment by master-still** [comment]
@HaoK 
Can you please provide guidance ?

![identity_workaround](https://cloud.githubusercontent.com/assets/2611892/22171941/6066392a-dfc2-11e6-9150-e6e4ec440b22.png)

I Must be missing something for my UserStore !

**Comment by VahidN** [comment]
It's fixed in v1.1: https://github.com/aspnet/Identity/issues/988

**Comment by master-still** [comment]
@VahidN Thank you

**Comment by MaklaCof** [comment]
If it helps, this is my UserStore:
```
    public class AppUserStore : UserStore<AppUser, AppRole, AppDbContext, int, AppUserClaim, AppUserRole, AppUserLogin, AppUserToken, AppRoleClaim>
    {
        public AppUserStore(AppDbContext context, IdentityErrorDescriber describer = null) : base(context, describer)
        {
        }

        protected override AppUserClaim CreateUserClaim(AppUser user, Claim claim)
        {
            return new AppUserClaim(user, claim);
        }

        protected override AppUserLogin CreateUserLogin(AppUser user, UserLoginInfo login)
        {
            return new AppUserLogin(user, login);
        }

        protected override AppUserRole CreateUserRole(AppUser user, AppRole role)
        {
            return new AppUserRole(user, role);
        }

        protected override AppUserToken CreateUserToken(AppUser user, string loginProvider, string name, string value)
        {
            return new AppUserToken(user, loginProvider, name, value);
        }
    }
```
And this is my AppRole:
```
    public partial class AppRole : IdentityRole<int, AppUserRole, AppRoleClaim>
    {
    }
```

**Comment by master-still** [comment]
@MaklaCof  https://github.com/master-still/test

currently I am having problem while passing my AppRole to Userstore.

**Comment by master-still** [comment]
Solved It ! Thank you all.

**Comment by mhannaford-oncore** [comment]
@master-still What did you do to fix this? I am using 1.1 and see the same error you reported. My code looks almost exactly the same as yours. Thanks

**Comment by mkontula** [comment]
Having the same issue. @master-still care to update your test repo with functioning code? Current version gives quite a bunch of build errors.

**Comment by Braed** [comment]
Can we PLEASE get some sort of official solution for this? MS Docs need updating.

**Comment by jeremyBass** [comment]
I hate to bring up what seems to be an old topic, but I have follow these threads, and all SO's on this, and still end up with the error

> Cannot instantiate implementation type 'WSU.Models.User.AppRoleStore' for service type 'Microsoft.AspNetCore.Identity.IRoleStore`1[WSU.Models.User.AppRoles]'.

Here are the basics I have going:

## *in Startup.cs*
```c#
            services.AddIdentity<AppUser, AppRoles>(options =>
            {
                options.Cookies.ApplicationCookie.SessionStore = new MemoryCacheTicketStore();
            })
            .AddUserStore<AppUserStore>()
            .AddUserManager<AppUserManager>()
            .AddRoleStore<AppRoleStore>()
            .AddRoleManager<AppRoleManager>()
            .AddSignInManager<AppSignInManager>()
            .AddDefaultTokenProviders();


            services.AddScoped<RoleStore<AppRoles, ApplicationDbContext, int, UserRole, RoleClaim>, AppRoleStore>();
            services.AddScoped<UserStore<AppUser, AppRoles, ApplicationDbContext, int, UserClaim, UserRole, UserLogin, UserToken, RoleClaim>, AppUserStore>();
            services.AddScoped<UserManager<AppUser>, AppUserManager>();
            services.AddScoped<RoleManager<AppRoles>, AppRoleManager>();
            services.AddScoped<SignInManager<AppUser>, AppSignInManager>();
            services.AddScoped<IUserStore<AppUser>, AppUserStore>();
            services.AddScoped<IRoleStore<AppRoles>, AppRoleStore>();


```


## *all store and services for the user*

```c#
   // Add profile data for application users by adding properties to the ApplicationUser class
    public class AppUser : IdentityUser<int, UserClaim, UserRole, UserLogin>
    {
        public bool pending_account { get; set; }
    }
    public class AppRoles : IdentityRole<int, UserRole, RoleClaim>
    {
        public AppRoles() { }
        public AppRoles(string role) : base(role) { }
        public string Description { get; set; }
        public DateTime CreatedDate { get; set; }
        public string IPAddress { get; set; }
    }

    public class UserRole : IdentityUserRole<int>
    {
        private AppRoles role;
        private AppUser user;

        //public UserRole() : base() { }

        public UserRole(AppUser user, AppRoles role)
        {
            this.user = user;
            this.role = role;
        }
    }
    public class UserClaim : IdentityUserClaim<int>
    {
        private Claim claim;
        private AppUser user;

        //public UserClaim() : base() { }

        public UserClaim(AppUser user, Claim claim)
        {
            this.user = user;
            this.claim = claim;
        }
    }
    public class UserLogin : IdentityUserLogin<int>
    {
        private UserLoginInfo login;
        private AppUser user;

        //public UserLogin() : base() { }
        public UserLogin(AppUser user, UserLoginInfo login)
        {
            this.user = user;
            this.login = login;
        }
    }
    public class UserToken : IdentityUserToken<int>
    {
        private AppUser user;
        //public UserToken() : base() { }
        public UserToken(AppUser user, string loginProvider, string name, string value)
        {
            this.user = user;
            LoginProvider = loginProvider;
            Name = name;
            Value = value;
        }
    }
    public class RoleClaim : IdentityRoleClaim<int>
    {
        // public RoleClaim() : base() { }
        private Claim claim;
        private AppRoles role;

        public RoleClaim(AppRoles role, Claim claim)
        {
            this.role = role;
            this.claim = claim;
        }
    }

    public class AppRoleManager : RoleManager<AppRoles>
    {
        public AppRoleManager(IRoleStore<AppRoles> store, IEnumerable<IRoleValidator<AppRoles>> roleValidators, ILookupNormalizer keyNormalizer, IdentityErrorDescriber errors, ILogger<RoleManager<AppRoles>> logger, IHttpContextAccessor contextAccessor) : base(store, roleValidators, keyNormalizer, errors, logger, contextAccessor)
        {
        }
    }

    public abstract class AppRoleStore : RoleStore<AppRoles, ApplicationDbContext, int, UserRole, RoleClaim> {
        public AppRoleStore(ApplicationDbContext context, IdentityErrorDescriber describer = null) : base(context, describer) { }

        protected override RoleClaim CreateRoleClaim(AppRoles role, Claim claim)
        {
            return new RoleClaim(role, claim);
        }
    }


    public class AppSignInManager : SignInManager<AppUser> {
        public AppSignInManager(AppUserManager userManager, IHttpContextAccessor contextAccessor, IUserClaimsPrincipalFactory<AppUser> claimsFactory, IOptions<IdentityOptions> optionsAccessor, ILogger<SignInManager<AppUser>> logger) : base(userManager, contextAccessor, claimsFactory, optionsAccessor, logger) { }
    }
    public class AppUserManager : UserManager<AppUser> {
        public AppUserManager(IUserStore<AppUser> store, IOptions<IdentityOptions> optionsAccessor, IPasswordHasher<AppUser> passwordHasher, IEnumerable<IUserValidator<AppUser>> userValidators, IEnumerable<IPasswordValidator<AppUser>> passwordValidators, ILookupNormalizer keyNormalizer, IdentityErrorDescriber errors, IServiceProvider services, ILogger<UserManager<AppUser>> logger) : base(store, optionsAccessor, passwordHasher, userValidators, passwordValidators, keyNormalizer, errors, services, logger) { }
    }

    public class AppUserStore : UserStore<AppUser, AppRoles, ApplicationDbContext, int, UserClaim, UserRole, UserLogin, UserToken, RoleClaim>
    {
        public AppUserStore(ApplicationDbContext context, IdentityErrorDescriber describer = null) : base(context, describer)
        {
        }

        protected override UserClaim CreateUserClaim(AppUser user, Claim claim)
        {
            return new UserClaim(user, claim);
        }

        protected override UserLogin CreateUserLogin(AppUser user, UserLoginInfo login)
        {
            return new UserLogin(user, login);
        }

        protected override UserRole CreateUserRole(AppUser user, AppRoles role)
        {
            return new UserRole(user, role);
        }

        protected override UserToken CreateUserToken(AppUser user, string loginProvider, string name, string value)
        {
            return new UserToken(user, loginProvider, name, value);
        }
    }

```


## And my DB context
```c#

    public class ApplicationDbContext : IdentityDbContext<AppUser, AppRoles, int, UserClaim, UserRole, UserLogin, RoleClaim, UserToken>
    {
        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
            : base(options)
        {
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {

            base.OnModelCreating(builder);

            builder.Entity<AppUser>()
                   .ToTable("Users")
                   .Property(p => p.Id)
                   .HasColumnName("user_id");

            builder.Entity<AppRoles>().ToTable("AppRoles");

            builder.Entity<UserRole>(i => {
                                                      i.ToTable("UserRole");
                                                      i.HasKey(x => new { x.RoleId, x.UserId });
                                                  });
            builder.Entity<UserLogin>(i => {
                i.ToTable("UserLogin");
                i.Property(e => e.UserId).HasColumnName("user_id");
                i.HasKey(x => new { x.ProviderKey, x.LoginProvider });
            });
            builder.Entity<RoleClaim>(i => {
                                                       i.ToTable("RoleClaims");
                                                       i.HasKey(x => x.Id);
                                                   });
            builder.Entity<UserClaim>(i => {
                                                       i.ToTable("UserClaims");
                                                       i.HasKey(x => x.Id);
                                                   });
            builder.Entity<UserToken>(i => {
                                                       i.ToTable("UserTokens");
                                                       i.HasKey(x => x.UserId);
                                                   });
        }

        public DbSet<AppUser> user { get; set; }
        public DbSet<AppRoles> roles { get; set; }
        public DbSet<RoleClaim> RoleClaims { get; set; }
        public DbSet<AppRoles> Roles { get; set; }
        public DbSet<UserClaim> UserClaims { get; set; }
        public DbSet<UserLogin> UserLogins { get; set; }
        public DbSet<UserRole> UserRoles { get; set; }
        public DbSet<UserToken> UserTokens { get; set; }
    }
```


I still end up with `Cannot instantiate implementation type 'WSU.Models.User.AppRoleStore' for service type 'Microsoft.AspNetCore.Identity.IRoleStore`1[WSU.Models.User.AppRoles]'.` as the error right out of the `Program.cs` as it launches.  Would anyone be able to help here?

**Comment by Cubelaster** [comment]
Had the same problem. My problem started when I have overriden the IdentityUser to 
`public class User : IdentityUser<Guid>`
This prompted the EF to not be able to build and show some wild errors. After finally figuring out that my DbContext needs to be in form of:
`public class NgSchoolsContext : IdentityDbContext<User, IdentityRole<Guid>, Guid>`
(notice the added IdentityRole<Guid> and Guid: why is it necessary that we explicitly say that IdentityRole is also a Guid? Shouldn't it derive from User as it's base entity? )
the Add-Migration wouldn't work and i got a "No parameterless constructor" error. That part was fixed by adding 
`services.AddIdentityCore<User>()
                .AddEntityFrameworkStores<NgSchoolsContext>()
                .AddRoles<IdentityRole<Guid>>()
                .AddDefaultTokenProviders();`
new Roles definition(?) in Startup. Also, why wouldn't it derive it's Id Type from it's IdentityUser implementation? 
After that the solution could be built again. However!! Add-Migration was working, but update scripts couldn't execute because the scripts tried to change all PK columns to another type (from string to Guid for me) and that is a no go. 
So after that, I dropped the Db and deleted all my migrations and just created a new initial migration and ran Update-Database. That solved it.
